### ISSUE TITLE
crash on all-day-event display  with "right" timezones

### ISSUE THREAD

**Issue by extmind** [pimutils/khal#330]
khal crashes on displaying the last day of an all-day-event that lasts for multiple days. The ics file says, that the event ends on 2016-02-09, but khal complains that 2016-02-08 is already out of range. I'm running khal 0.7.0 on Arch Linux using [the corresponding AUR package](https://aur.archlinux.org/packages/khal/).

Steps to reproduce:
[I've attached the event's .ics file (renamed to .txt to make GitHub happy...).](https://github.com/geier/khal/files/111579/8e29c4b5-a6f7-431d-a7a6-33e5e0f66c97.ics.txt)
Today is 2016-02-01. When I run `khal calendar --days 7` it works as expected. But when I run `kahl calendar --days 8` it does not. The resulting error is:

```
Traceback (most recent call last):
  File "/usr/bin/khal", line 9, in <module>
    load_entry_point('khal==0.7.0', 'console_scripts', 'khal')()
  File "/usr/lib/python3.5/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3.5/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3.5/site-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3.5/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3.5/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/usr/lib/python3.5/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/lib/python3.5/site-packages/khal/cli.py", line 254, in calendar
    full=full
  File "/usr/lib/python3.5/site-packages/khal/controllers.py", line 149, in calendar
    show_all_days=show_all_days, full=full, **kwargs)
  File "/usr/lib/python3.5/site-packages/khal/controllers.py", line 122, in get_agenda
    items = event.relative_to(day, full).splitlines()
  File "/usr/lib/python3.5/site-packages/khal/khalendar/event.py", line 531, in relative_to
    end=self.end)))
ValueError: Day out of range: {'day': datetime.date(2016, 2, 8), 'start': datetime.date(2016, 2, 3), 'end': datetime.date(2016, 2, 7)}
```

**Comment by hobarrera** [comment]
I can't reproduce this with the latest version from master. Can you confirm if you have the same issue? (you can use `khal-git` from AUR to try this quickly).

**Comment by extmind** [comment]
I've installed khal-git from AUR and I still get about the same error.
```
Traceback (most recent call last):
  File "/usr/bin/khal", line 9, in <module>
    load_entry_point('khal==0.7.1.dev103+ng6bcd842.d20160201', 'console_scripts', 'khal')()
  File "/usr/lib/python3.5/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3.5/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3.5/site-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3.5/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3.5/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/usr/lib/python3.5/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/lib/python3.5/site-packages/khal/cli.py", line 255, in calendar
    bold_for_light_color=ctx.obj['conf']['view']['bold_for_light_color']
  File "/usr/lib/python3.5/site-packages/khal/controllers.py", line 147, in calendar
    **kwargs)
  File "/usr/lib/python3.5/site-packages/khal/controllers.py", line 115, in get_agenda
    items = event.relative_to(day, full).splitlines()
  File "/usr/lib/python3.5/site-packages/khal/khalendar/event.py", line 556, in relative_to
    end=self.end)))
ValueError: Day out of range: {'day': datetime.date(2016, 2, 8), 'end': datetime.date(2016, 2, 7), 'start': datetime.date(2016, 2, 3)}
```

**Comment by hobarrera** [comment]
Can you share your `khal.conf`, please?

**Comment by geier** [comment]
What is your `local_timezone` and `default_timezone` set to? (see output of khal -v). Also, please try to delete your local khal.db and try again.

**Comment by extmind** [comment]
My locale settings according to khal -v:
```
debug: [locale]
debug:   local_timezone: Europe/Berlin
debug:   default_timezone: Europe/Berlin
debug:   timeformat: %H:%M
debug:   dateformat: %d.%m.
debug:   longdateformat: %d.%m.%Y
debug:   datetimeformat: %d.%m. %H:%M
debug:   longdatetimeformat: %d.%m.%Y %H:%M
debug:   firstweekday: 0
debug:   encoding: utf-8
debug:   unicode_symbols: True
debug:   weeknumbers: False
```

After I installed the khal-git version, I had to delete khal.db before issuing another command. (Also when I downgraded to 0.7.0 afterwards.)

**Comment by geier** [comment]
Strangely, I cannot reproduce this currently, but I haven't forgotten about this yet.

**Comment by extmind** [comment]
I probably found out why this is difficult to reproduce. This bug occurs exactly when the operating system timezone is set to a leap-second accounting timezone, i.e. `/usr/share/zoneinfo/right/*`. As this is an uncommon thing to do in the first place, it is not a problem for me if this bug is set to wontfix.

**Comment by geier** [comment]
Thank you for tracking this down. I'll need to look into this and if it can be supported with a reasonable amount of work. Do you have both (default and local) timezones set to a timezone from `right`?

What are the (practical) advantages using a TAI timezone in the first place?

**Comment by extmind** [comment]
> Do you have both (default and local) timezones set to a timezone from `right`?
Both settings in khal are set to Europe/Berlin. If I try to set one of these to right/Europe/Berlin a config error says that the given timezone is unknown.
Instead the system's timezone setting was set to right/Europe/Berlin.

**Comment by geier** [comment]
I have actually set my desktop to `right` timezones and khal is working fine for me. I believe this bug has been fixed with the backend rewrite prior to v0.8.

I will therefore close this issue, please re-open or comment if you can still reproduce this issue.
